internal/promotion/runner/builtin/tar_extractor.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	return t.run(ctx, stepCtx, cfg)
+}
+
+func (t *tarExtractor) run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+	cfg builtin.UntarConfig,
+) (promotion.StepResult, error) {
+	// Secure join the paths to prevent path traversal attacks.
+	inPath, err := securejoin.SecureJoin(stepCtx.WorkDir, cfg.InPath)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not secure join inPath %q: %w", cfg.InPath, err)
+	}
+	outPath, err := securejoin.SecureJoin(stepCtx.WorkDir, cfg.OutPath)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not secure join outPath %q: %w", cfg.OutPath, err)
+	}
+
+	// Ensure the output directory exists
+	if err := os.MkdirAll(outPath, 0755); err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("failed to create output directory %q: %w", cfg.OutPath, err)
+	}
+
+	// Load the ignore rules.
+	matcher, err := t.loadIgnoreRules(outPath, cfg.Ignore)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("failed to load ignore rules: %w", err)
+	}
+
+	// Open the tar file
+	file, err := os.Open(inPath)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("failed to open tar file %q: %w", cfg.InPath, err)
+	}
+	defer file.Close()
+
+	logger := logging.LoggerFromContext(ctx)
+	var tarReader *tar.Reader
+
+	// Check if the file is gzipped
+	gzr, err := gzip.NewReader(file)
+	if err == nil {
+		// File is gzipped
+		defer gzr.Close()
+		tarReader = tar.NewReader(gzr)
+		logger.Debug("treating file as gzipped tar")
+	} else {
+		// File is not gzipped, reset to beginning of file and try as regular tar
+		_, err = file.Seek(0, 0)
+		if err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("failed to seek in tar file: %w", err)
+		}
+		tarReader = tar.NewReader(file)
+		logger.Debug("treating file as regular tar")
+	}
+
+	// Extract the tar file
+	stripComponents := int64(0)
+	if cfg.StripComponents != nil {
+		stripComponents = *cfg.StripComponents
+		if stripComponents > 0 {
+			logger.Debug("stripping path components", "count", stripComponents)
+		}
+	}
+
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+		if err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("error reading tar: %w", err)
+		}
+
+		// Skip if this is not a regular file, symlink, or directory
+		if header.Typeflag != tar.TypeReg && header.Typeflag != tar.TypeDir && header.Typeflag != tar.TypeSymlink {
+			logger.Debug("skipping non-regular file", "path", header.Name, "type", header.Typeflag)
+			continue
+		}
+
+		// Handle stripping components if specified
+		targetName := header.Name
```

Comment:
Further validation is required to ensure it does not attempt to escape the working directory, which can happen with a maliciously crafted compressed artifact. E.g. something like:

```
func validRelPath(p string) bool {
	if p == "" || strings.Contains(p, `\`) || strings.HasPrefix(p, "/") || strings.Contains(p, "../") {
		return false
	}
	return true
}
```